Fix for #3373 Changed next time stamp implementation.

kobalski opened the pull request:
As discussed in https://github.com/hazelcast/hazelcast/issues/3068 before, our time stamp implementation used by hibernate query cache is millisecond based which is taken with instance.getCluster().getClusterTime(). It causes test failures in fast machines form time to time. I have modified nextTimestamp(HazelcastInstance instance) so that it waits until a different new timestamp value is generated. This implementation is similar to ehcache implementation which is considered as reference implementation for hibernate second level cache.

mdogan commented on hazelcast-hibernate/hazelcast-hibernate3/src/main/java/com/hazelcast/hibernate/HazelcastTimestamper.java:
```diff
@@ -29,13 +31,27 @@
     private HazelcastTimestamper() {
     }
 
     public static long nextTimestamp(HazelcastInstance instance) {
```
Can you add a documentation over the method describing the algorithm used here, what it does, what it guarantees, why it's not just a regular timestamp like `System.currentTimeMillis()`, what `BIN_DIGITS` means, what's stored in `VALUE` .. etc?

mdogan commented:
Is it really 
>it waits until a different new timestamp value is generated

or more like *generates a different timestamp on each call*?

kobalski commented:
It waits in the while loop part, but what it does is basically as you said, generating different time stamp. VALUE field is for making sure every value method returns is different. I changed BIN_DIGITS field name to BASE_TIME_SHIFT and added some comments and logs to the code.

mesutcelik commented:
checkstyle failures. execute following to check them locally after the fix

```
mvn clean validate -P checkstyle
```

ahmetmircik commented on hazelcast-hibernate/hazelcast-hibernate3/src/main/java/com/hazelcast/hibernate/HazelcastTimestamper.java:
```diff
@@ -29,13 +32,42 @@
-        return instance.getCluster().getClusterTime();
+        int runs = 0;
+        while (true) {
+            long base = instance.getCluster().getClusterTime() << BASE_TIME_SHIFT;
```
If negative value for `base` causes problems, this line makes the code fragile. It seems better to me to put a check if a negative `base` is an undesired thing.

kobalski replied:
```
 for (long current = VALUE.get(), update = Math.max(base, current + 1); update < maxValue;
        current = VALUE.get(), update = Math.max(base, current + 1)) {
```
In the for loop above max value between base and current + 1 is chosen, if negative value occurs due to bit shifting it will not effect the return value.

ahmetmircik replied:
But what if also `maxValue` is negative, don't you end up an eternal while loop?

kobalski replied:
Yes you are right, it may cause an infinite loop scenario, however inside the for loop, if statement will return the value no matter what,
```
 if (VALUE.compareAndSet(current, update)) {
         if (runs > 1) {
               Logger.getLogger(HazelcastTimestamper.class)
                    .finest(String.format("Waiting for time to pass. Looped %d times", runs));
          }
           return update;
  }
```
because update = Math.max(base, current + 1) will assign update to current + 1 and method will return inside if statement because VALUE.compareAndSet(current, update) will return true because update will be current + 1. Again, you are right the loop can be constructed as an infinite loop because update < maxValue will never be satisfied, however the method will always return in the first run of the loop.

ahmetmircik commented on hazelcast-hibernate/hazelcast-hibernate4/src/main/java/com/hazelcast/hibernate/HazelcastTimestamper.java:
```diff
@@ -29,20 +32,49 @@
             if (cfg.getTimeToLiveSeconds() > 0) {
-                // TTL in ms
+                // TTL in ms.
                 return cfg.getTimeToLiveSeconds() * SEC_TO_MS;
```
This line is not related to this PR but it may suffer from an integer overflow.

kobalski commented:
Also added integer overflow check for ``` public static int getTimeout(HazelcastInstance instance, String regionName) ``` in order to prevent returning negative value for timeout.

ahmetmircik commented:
verify

mesutcelik commented:
@kobalski ,

I am closing that right now. There is another fix attempt at 43966bf and 4f4c09d9cc18638b36bc80c8824ce946cd64408d
